### ISSUE TITLE
Consolidate CSS stylesheets

### ISSUE THREAD

**Issue by huyenltnguyen** [freeCodeCamp/ui#61]
This ticket is semi-related to https://github.com/freeCodeCamp/freeCodeCamp/issues/52030. 

In order to replace Bootstrap in /learn, we need to export a stylesheet from the component library to be used as the `bootstrap.min.css` replacement. However, the stylesheet situation in the component library itself needs to be sorted out, and this ticket is to track this work.

## Current problem

The library currently has multiple stylesheets. Their purposes aren't clear, they have many similar CSS rules and some of the CSS rules don't match, which could cause styling issue if the import order changes.

The stylesheets are:
- https://github.com/freeCodeCamp/ui/blob/main/src/colors.css
  - Contains color tokens
- https://github.com/freeCodeCamp/ui/blob/main/src/base.css: 
  - Contains Tailwind styles as well as some Tailwind overrides
  - Was created as part of the Tailwind setup, following this documentation: https://tailwindcss.com/docs/installation/using-postcss
- https://github.com/freeCodeCamp/ui/blob/main/src/global-element-styles.css:
  - Is a subset of https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.css, containing element styles (styles that use HTML element selectors and affect elements directly)
    https://github.com/freeCodeCamp/ui/blob/3886b615b6cb57840295fddc0c08a8d211e9ff3a/src/global-element-styles.css#L1-L3
- https://github.com/freeCodeCamp/ui/blob/main/src/normalize.css
  - Is used for resetting default browser styles
  - Based on a comment in the file, the code was copied from https://necolas.github.io/normalize.css/3.0.3/normalize.css

Observation:
- Except for `colors.css`, the stylesheets overlap each other:
  - https://github.com/freeCodeCamp/ui/blob/3886b615b6cb57840295fddc0c08a8d211e9ff3a/src/base.css#L16-L18
  - https://github.com/freeCodeCamp/ui/blob/3886b615b6cb57840295fddc0c08a8d211e9ff3a/src/global-element-styles.css#L397-L400
  - https://github.com/freeCodeCamp/ui/blob/3886b615b6cb57840295fddc0c08a8d211e9ff3a/src/normalize.css#L40-L42
  - (Not to mention that we do have custom CSS in the implementation of each component, in the form of class names, and those custom CSS rules override the base)
- The stylesheets also have some conflicting rules:
  - https://github.com/freeCodeCamp/ui/blob/3886b615b6cb57840295fddc0c08a8d211e9ff3a/src/normalize.css#L61-L64
  - https://github.com/freeCodeCamp/ui/blob/3886b615b6cb57840295fddc0c08a8d211e9ff3a/src/global-element-styles.css#L150-L155
- `normalize.css` is supposed to match the original https://necolas.github.io/normalize.css/3.0.3/normalize.css, but I found some discrepancies: https://github.com/huyenltnguyen/freeCodeCamp-ui/pull/1
- Tailwind provides a [preflight stylesheet](https://tailwindcss.com/docs/preflight) out of the box, and the purpose of the stylesheet is similar to `normalize.css`. However, we turned this option off in favor of our custom stylesheets (`global-element-styles.css` and `normalize.css`)
  - https://github.com/freeCodeCamp/ui/blob/1e3252fb0497b1b3ec29e0ddb1ff9a51fc908589/tailwind.config.js#L6
- From what I understand, we added `global-element-styles.css` and `normalize.css` in because we wanted to mimic the /learn environment in Storybook, in order to develop components for /learn
   - https://github.com/freeCodeCamp/ui/blob/3886b615b6cb57840295fddc0c08a8d211e9ff3a/.storybook/preview.js#L1-L5
   - According to the import order, if there is a conflict between `global-element-styles.css` and `normalize.css`, `global-element-styles.css` will take precedence.
- Only `colors.css` and `base.css` are bundled
  - https://github.com/freeCodeCamp/ui/blob/3886b615b6cb57840295fddc0c08a8d211e9ff3a/src/base.css#L1
  - https://github.com/freeCodeCamp/ui/blob/4dfc72ecd15bc6fce0826436ea2eeaab1c1e2c92/package.json#L32

## What we want to achieve

For the component library itself, we want:
  - A simplified styling strategy

For the library consumers, we want:
  - The color variables
  - A stylesheet for resetting default browser styles. This should be the same stylesheet that the component library uses internally

## Approaches

I think we should start developing the component library as a standalone package, rather than having /learn influence it.

So I'd propose:
- Removing `global-element-styles.css` since this comes from Bootstrap, and includes a lot of CSS customization, not just reset
- Using Tailwind preflight for browser styles reset (this means simply flipping the `preflight` flag to `true`)
  - Here is a comparison between `normalize.css` and Tailwind preflight stylesheets: [huyenltnguyen/freeCodeCamp-ui@`f81770d` (#2)](https://github.com/huyenltnguyen/freeCodeCamp-ui/pull/2/commits/f81770d686b944262569a724d1c8bc432a2e23b7)
  - The differences between them aren't significant. But my concern with `normalize.css` is the library is not actively maintained, while Tailwind is an active project and its preflight stylesheet is written upon [modern-normalize](https://github.com/sindresorhus/modern-normalize), which is... more modern.
  - We are already using Tailwind, so I think snapping back to Tailwind preflight would prevent stylesheets conflicts, and allow us to remove an additional dependency
- Removing `normalize.css`, if we choose to use Tailwind preflight

As for the styling strategy in the component library, I think:
- Each component should be styled using Tailwind classes. This should be the first option when it comes to styling a component.
- If we want to override Tailwind globally, the rules should be added to `base.css`. As the file name suggests, the file should house very basic CSS styles, and we want to be mindful when adding changes as they will require overrides on the component level, which would bring inconsistencies.

## Implications for the Bootstrap removal in /learn

Given that we tried replacing the Bootstrap stylesheet with `normalize.css` and the result was positive ([ref](https://github.com/freeCodeCamp/freeCodeCamp/issues/52030#issuecomment-2039234396)), I don't think the Tailwind preflight would cause a catastrophe.

/learn has a huge [global.css](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/components/layouts/global.css) file that has all sort of CSS rules to override Bootstrap. I think those rules would help keep the UI intact (or mostly intact) when we swap the stylesheets. (The `global.css` file also needs a round of review and cleanup as part of the Bootstrap removal, but that's a separate topic.)

**Comment by ahmaxed** [comment]
It does not have to be my birthday. I always celebrate when I see well-thought out issues like this.
A few thoughts:

1. At the time, I came to a conclusion that normalize is the industry standard comparing to preflight and it would make removing bootstrap easier since bootstrap was using it. If any of these assumptions are incorrect, we should move forward with preflight.

2. There should be one set of presets, for the component library regardless of what is being overwritten in learn, etc.

3. I think the main reason we added normalize and global element styles is that when the elements were imported to learn they would look different from Storybook. Might need to remove them, and see how they effect storybook and learn.

Finally, this is a great blueprint for the stylesheet consolidation. At the same time, it might be helpful to evaluate how it would impact the migration. If we decide to do this before or after the migration, I am on board either way. It is a task that we need to take care of at some point.
